src/Npgsql/NpgsqlSchema.cs

```diff
@@ -339,27 +424,43 @@ pg_catalog.pg_class i
 WHERE
     i.relkind = 'i' AND
     n.nspname NOT IN ('pg_catalog', 'pg_toast') AND
-    t.relkind = 'r'");
-
-        using var command = BuildCommand(conn, getIndexes, restrictions, false, "current_database()", "n.nspname", "t.relname", "i.relname");
-        using var adapter = new NpgsqlDataAdapter(command);
-        await adapter.Fill(indexes, async, cancellationToken).ConfigureAwait(false);
-
-        return indexes;
+    t.relkind = 'r'
+""");
+
+        return ParseResults(
+            async,
+            BuildCommand(conn, sql, restrictions, false, "current_database()", "n.nspname", "t.relname", "i.relname"),
+            dataTable,
+            (reader, row) =>
+            {
+                row["table_catalog"] = GetFieldValueOrDBNull<string>(reader, 0);
+                row["table_schema"] = GetFieldValueOrDBNull<string>(reader, 1);
+                row["table_name"] = GetFieldValueOrDBNull<string>(reader, 2);
+                row["index_name"] = GetFieldValueOrDBNull<string>(reader, 3);
+                row["type_desc"] = GetFieldValueOrDBNull<string>(reader, 4);
+            }, cancellationToken);
     }
 
-    [RequiresUnreferencedCode("Members from serialized types or types used in expressions may be trimmed if not referenced directly.")]
-    static async Task<DataTable> GetIndexColumns(NpgsqlConnection conn, string?[]? restrictions, bool async, CancellationToken cancellationToken = default)
+    static Task<DataTable> GetIndexColumns(NpgsqlConnection conn, string?[]? restrictions, bool async, CancellationToken cancellationToken = default)
     {
-        var indexColumns = new DataTable("IndexColumns") { Locale = CultureInfo.InvariantCulture };
-
-        indexColumns.Columns.AddRange(new[] {
-            new DataColumn("constraint_catalog"), new DataColumn("constraint_schema"), new DataColumn("constraint_name"),
-            new DataColumn("table_catalog"), new DataColumn("table_schema"), new DataColumn("table_name"),
-            new DataColumn("column_name"), new DataColumn("index_name")
-        });
+        var dataTable = new DataTable("IndexColumns")
+        {
+            Locale = CultureInfo.InvariantCulture,
+            Columns =
+            {
+                new DataColumn("constraint_catalog"),
+                new DataColumn("constraint_schema"),
+                new DataColumn("constraint_name"),
+                new DataColumn("table_catalog"),
+                new DataColumn("table_schema"),
+                new DataColumn("table_name"),
+                new DataColumn("column_name"),
+                new DataColumn("index_name")
+            }
+        };
 
-        var getIndexColumns = new StringBuilder(@"
+        var sql = new StringBuilder(
+            """
 SELECT
     current_database() AS constraint_catalog,
     t_ns.nspname AS constraint_schema,
```

Comment:
Can optionally remove all the aliases, since reading is now fully ordinal based.